<a href="https://colab.research.google.com/github/JulioRan09/Porgramacion-Predictiva/blob/main/Actividad_Calificada_07_Datos_Duplicados_MATRICULA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | 25 Septiembre |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | Julio Rangel García |
| **Matrícula** | 276572 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [2]:
# Escribe tu código aquí
#hoteles_count = hoteles.copy()

hoteles_first = hoteles.duplicated(keep='first').sum()
print(f'Las filas duplicadas utilizando "keep=first" son: {hoteles_first}')

hoteles_last = hoteles.duplicated(keep='last').sum()
print(f'Las filas duplicadas utilizando "keep=last" son: {hoteles_last}')

hoteles_false = hoteles.duplicated(keep=False).sum()
print(f'Las filas duplicadas utilizando "keep=False" son: {hoteles_false}')

Las filas duplicadas utilizando "keep=first" son: 31994
Las filas duplicadas utilizando "keep=last" son: 31994
Las filas duplicadas utilizando "keep=False" son: 40165


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [3]:
# Escribe tu código aquí
#Si contaramos con conservar un dato de los duplicados (ya sea con first o last), podemos calcular el porcentaje de filas sobrantes como
#el total de filas menos las filas de duplicados / total de filas

total_filas = len(hoteles)
filas_sobrantes = total_filas - hoteles_first
porcentaje_filas_sobrantes = (filas_sobrantes / total_filas) * 100
print(f'El porcentaje de filas sobrantes respecto al total es: {porcentaje_filas_sobrantes:.2f}%')

El porcentaje de filas sobrantes respecto al total es: 73.20%


In [4]:
#Utilizamos el value counts para agrupar los repetidos y ponemos la condición de que aparezca más de una vez para ver cuántos elementos tenemos por grupo.

hoteles_repetidos = hoteles.value_counts()
repetidos = hoteles_repetidos[hoteles_repetidos > 1]
len_repetidos = len(repetidos)


print('Los grupos repetidos son:')
print(repetidos)
print(f'Hay un total de {len_repetidos} grupos repetidos')

Los grupos repetidos son:
hotel         is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type    adr     required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel    0            256        2016               October             43                        16                         2                        3                     2       0.0       0       BB    DEU      Online TA       TA/TO                 0                  0                       0                               A                   A                   0                No Deposit    9.0    

### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [5]:
# Escribe tu código aquí
hoteles_duplicados = hoteles.copy()
hoteles_duplicados['es_duplicado'] = hoteles_duplicados.duplicated(keep='first')
hoteles_duplicados_por_hotel = hoteles_duplicados.groupby('hotel').sum()
hoteles_duplicados_por_hotel['porcentaje'] = (hoteles_duplicados_por_hotel['es_duplicado'] / hoteles_duplicados.shape[0]) * 100


hoteles_duplicados_por_hotel

,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,...,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,es_duplicado,porcentaje
hotel,,,,,,,,,,,,,,,,,,,,,
City Hotel,33102,8705335,159943106,JulyJulyJulyJulyJulyJulyJulyJulyJulyJulyJulyJu...,2155987,1252353,63082,173174,146838,7248.0,...,535907.0,255980,TransientTransientTransientTransientTransientT...,8353803.24,1933,43387,Check-OutCanceledCanceledCanceledCanceledCance...,2015-07-032015-07-012015-04-302015-06-232015-0...,25902,21.695284
Resort Hotel,11122,3712588,80765825,JulyJulyJulyJulyJulyJulyJulyJulyJulyJulyJulyJu...,1087263,633799,47664,125337,74798,5155.0,...,750539.0,21142,TransientTransientTransientTransientTransientT...,3803814.36,5531,24828,Check-OutCheck-OutCheck-OutCheck-OutCheck-OutC...,2015-07-012015-07-012015-07-022015-07-022015-0...,6092,5.102605


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

_Como podemos observar de la coumna porcentaje en el dataframe Hoteles_duplicados_por hotel, observamos que los duplicados afectan más para el CityHotel con un 21.70% de datos duplucados, mientras que en el ResortHotel apenas se alcanza u 5.10% de datos duplicados. Adicionalmente, observamos que estos datos se confirman mediante el análisis del paso 1.2 con un porcentaje total de datos duplicados de 26.80%._

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [6]:
#Primero creamos dos listas, una para guardar las columnas que mostraremos al final del ejercicio
#En la segunda lista guardaré los que considero los criterios más útiles para ordenar los datos repetidos

columnas_inspeccion = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'adults', 'country', 'market_segment', 'deposit_type', 'adr', 'reservation_status']
columnas_orden = ['hotel', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'country', 'lead_time', 'adr']

duplicados_completos = hoteles[hoteles.duplicated(keep=False)]
#Utilizamos la lista de columnas orden que creamos para ordenarlas por estos criterios
duplicados_ordenados = duplicados_completos.sort_values(by=columnas_orden)


#Mostramos los primeros 12 renglones
display(duplicados_ordenados[columnas_inspeccion].head(12))

,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40574,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40580,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40588,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40562,City Hotel,117,2015,August,1,1,ITA,Offline TA/TO,No Deposit,72.5,Check-Out
40585,City Hotel,117,2015,August,1,1,ITA,Offline TA/TO,No Deposit,72.5,Check-Out
40563,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40564,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40566,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40567,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40568,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [7]:
# Escribe tu código aquí
conteo_combinaciones = hoteles.value_counts(dropna=False)
grupos_repetidos = conteo_combinaciones[conteo_combinaciones > 1]
tamanos_grupos = grupos_repetidos.value_counts().sort_index(ascending=True)

print("Filas repetidas |", tamanos_grupos.head(10))

#Esto nos muestra cuantas filas tiene repetidas y la frecuencia con que este patrón aparece
#Por ejemplo, 3 filas repetidas aparece un total de 969 veces.


Filas repetidas | count
2     5034
3      969
4      497
5      275
6      171
7      115
8      103
9       93
10     123
11      60
Name: count, dtype: int64


Ahora encontramos el valor que más se repite con la instrucción .max() o bien sorteando el value counts anterior con ascending=False. De ambas maneras obtenemos que hay una fila que se repite 180 veces. Guardamos esta fila en combiación_max y la mostramos para darnos cuenta qué datos contiene.

In [8]:
grupo_max = grupos_repetidos.max()
combinacion_max = grupos_repetidos[grupos_repetidos == grupo_max]
print(combinacion_max)

hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel  1            277        2016               November            46                        7                          1                        2                     2       0.0       0       BB    PRT      Groups          TA/TO                 0                  0                       0                               A                   A                   0                Non Refund    NaN    NaN      0                     Tr

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [9]:
# Escribe tu código aquí
# Filtrar las filas duplicadas (todas las instancias que son parte de un grupo duplicado)
duplicados23 = hoteles[hoteles.duplicated(keep=False)]
hoteles_proporciones = hoteles.copy()


# Distribución de 'market_segment'
distribucion_market_segment = hoteles_proporciones['market_segment'].value_counts(normalize=True)

# Distribución de 'market_segment' solo duplicados
b_distribucion_market_segment = duplicados23['market_segment'].value_counts(normalize=True)

print("Proporciones de market_segment para todo el data set")
print(distribucion_market_segment)
print("\n==================================================")
print("\nProporciones de market_segmet sólo para los datos duplicados")
print(b_distribucion_market_segment)


Proporciones de market_segment para todo el data set
market_segment
Online TA        0.473046
Offline TA/TO    0.202856
Groups           0.165935
Direct           0.105587
Corporate        0.044350
Complementary    0.006223
Aviation         0.001985
Undefined        0.000017
Name: proportion, dtype: float64


Proporciones de market_segmet sólo para los datos duplicados
market_segment
Groups           0.416109
Offline TA/TO    0.303996
Online TA        0.205876
Corporate        0.036699
Direct           0.034981
Complementary    0.001867
Aviation         0.000473
Name: proportion, dtype: float64


In [10]:
# Distribución de 'deposit_type'
distribucion_deposit_type = hoteles_proporciones['deposit_type'].value_counts(normalize=True)

# Distribución de 'deposit_type' sólo duplicados
b_distribucion_deposit_type = duplicados23['deposit_type'].value_counts(normalize=True)



print("Proporciones de Deposit Type para todo el data set")
print(distribucion_deposit_type)
print("\n==================================================")
print("\nProporciones de Deposit Type sólo para los datos duplicados")
print(b_distribucion_deposit_type)


Proporciones de Deposit Type para todo el data set
deposit_type
No Deposit    0.876464
Non Refund    0.122179
Refundable    0.001357
Name: proportion, dtype: float64


Proporciones de Deposit Type sólo para los datos duplicados
deposit_type
No Deposit    0.641678
Non Refund    0.356430
Refundable    0.001892
Name: proportion, dtype: float64


In [11]:
# Distribución 'is_canceled'
distribucion_is_canceled = hoteles_proporciones['is_canceled'].value_counts(normalize=True)

# Distribución 'is_canceled' sólo duplicados
b_distribucion_is_canceled = duplicados23['is_canceled'].value_counts(normalize=True)

print("Proporciones de is_canceled para todo el data set")
print(distribucion_is_canceled)
print("\n==================================================")
print("\nProporciones de is_canceled sólo para los datos duplicados")
print(b_distribucion_is_canceled)

Proporciones de is_canceled para todo el data set
is_canceled
0    0.629584
1    0.370416
Name: proportion, dtype: float64


Proporciones de is_canceled sólo para los datos duplicados
is_canceled
1    0.583867
0    0.416133
Name: proportion, dtype: float64


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

_a) Basándonos en columnas temporales, como lead time, fechas de llegadas y confirmando con dátos específicos como cantidades de personas, de gasto, tipo de pagos, es muy difícil NO concluir que se trate de errores de registro._
_Podemos ser flexibles y definir 2 filas duplicadas como un umbral de cantidad de filas duplicadas para el cuál no los tratemos como error de registro si no como reservas distintas que coincidan en todas sus entradas, sin embargo, es poco confiable ya que coincidir en 32 entradas pareciera no ser posible._


_b) 'duplicated()' en 'pandas' trata los valores 'NaN' (Not a Number) como valores iguales a efectos de detección de duplicados. Es decir, si dos filas tienen 'NaN' en la misma columna, 'duplicated()' las considerará idénticas en esa columna. Esto es el comportamiento por defecto y es importante para este análisis._

_La implicación es que si dos reservas son idénticas en todas las demás columnas y ambas tienen 'NaN' tanto en 'agent' como en 'company' (lo cual es muy común en este dataset), 'duplicated()' las marcará como duplicadas. Esto es deseable para el propósito de este análisis, ya que un 'NaN' en 'agent' o 'company' significa que no hay un agente o compañía asociada a esa reserva. Si dos reservas tienen exactamente la misma falta de información en estas columnas (y son idénticas en todo lo demás), deben considerarse duplicados exactos._

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [12]:
# Escribe tu código aquí
# Número de filas antes de eliminar duplicados
filas_antes = hoteles.shape[0]
print(f"Número de filas antes de eliminar duplicados: {filas_antes}")

# Crear hoteles_limpio eliminando los duplicados exactos, conservando la primera aparición
hoteles_limpio = hoteles.drop_duplicates(keep='first')

# Renumerar el índice
hoteles_limpio = hoteles_limpio.reset_index(drop=True)

# Número de filas después de eliminar duplicados
filas_despues = hoteles_limpio.shape[0]
print(f"Número de filas después de eliminar duplicados: {filas_despues}")

# Cuántas filas se eliminaron
filas_eliminadas = filas_antes - filas_despues
print(f"Número de filas eliminadas: {filas_eliminadas}")

Número de filas antes de eliminar duplicados: 119390
Número de filas después de eliminar duplicados: 87396
Número de filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [13]:
# Escribe tu código aquí
# Para el DataFrame original (hoteles)
resumen_original = hoteles.groupby('hotel').agg(
    num_reservas=('hotel', 'size'),
    tasa_cancelacion=('is_canceled', 'mean'),
    adr_promedio=('adr', 'mean')
).reset_index()

print("Resumen de métricas para el DataFrame original:\n")
display(resumen_original)

print("\n" + "==============================" + "\n")

# Para el DataFrame sin duplicados (hoteles_limpio)
resumen_limpio = hoteles_limpio.groupby('hotel').agg(
    num_reservas=('hotel', 'size'),
    tasa_cancelacion=('is_canceled', 'mean'),
    adr_promedio=('adr', 'mean')
).reset_index()

print("Resumen de métricas para el DataFrame limpio (sin duplicados):\n")
display(resumen_limpio)

Resumen de métricas para el DataFrame original:



,hotel,num_reservas,tasa_cancelacion,adr_promedio
0,City Hotel,79330,0.417270,105.304465
1,Resort Hotel,40060,0.277634,94.952930




Resumen de métricas para el DataFrame limpio (sin duplicados):



,hotel,num_reservas,tasa_cancelacion,adr_promedio
0,City Hotel,53428,0.300386,110.985944
1,Resort Hotel,33968,0.234809,99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [14]:
# Escribe tu código aquí
# Agregar la columna 'version' a cada DataFrame resumen
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

# Unir los DataFrames con pd.concat()
comparacion = pd.concat([resumen_original, resumen_limpio])

# Renumerar el índice y ordenar por 'hotel'
comparacion = comparacion.reset_index(drop=True).sort_values(by='hotel')

print("Tabla comparativa de métricas por hotel:\n")
display(comparacion)

Tabla comparativa de métricas por hotel:



,hotel,num_reservas,tasa_cancelacion,adr_promedio,version
0,City Hotel,79330,0.417270,105.304465,original
2,City Hotel,53428,0.300386,110.985944,sin duplicados
1,Resort Hotel,40060,0.277634,94.952930,original
3,Resort Hotel,33968,0.234809,99.025346,sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

_a) La métrica que más cambia es el número de reservas, como se concluyó desde la Actividad 2, el cambio se esperaba mayor para el City Hotel, donde el número de reservas disminuyeron en aproximadamente 24k reservas, mientras que para el Resort Hotel se eliminaron aproximadamente 6k reservas._
_En esta misma línea de análisis, la tasa de cancelación se redujo un 10% para el city hotel y 4% para el Resort Hotel cuando eliminamos los dato duplicados, esta es una métrica importante para entrenar modelos predictivos._


_b) La presencia de registros duplicados podría sesgar el entrenamiento del modelo, ya que algunas reservas tendrían más peso que otras en el aprendizaje. En este conjunto de datos, los duplicados parecen aumentar artificialmente la proporción de cancelaciones, por lo que el modelo podría aprender una tasa de cancelación superior a la real. Además, si registros duplicados aparecen simultáneamente en los conjuntos de entrenamiento y prueba, las métricas de evaluación podrían verse infladas y dar una impresión demasiado optimista del rendimiento del modelo._

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [15]:
# Escribe tu código aquí
# 1. Verificar que no quedan duplicados exactos
num_duplicados_limpio = hoteles_limpio.duplicated().sum()
print(f"Número de duplicados exactos en hoteles_limpio: {num_duplicados_limpio}")

# 2. Verificar que no hay índices repetidos y el índice va de 0 a len(hoteles_limpio) - 1
indice_correcto = (hoteles_limpio.index == range(len(hoteles_limpio))).all()
print(f"El índice está correctamente renumerado y no tiene repetidos: {indice_correcto}")

Número de duplicados exactos en hoteles_limpio: 0
El índice está correctamente renumerado y no tiene repetidos: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [16]:
# Escribe tu código aquí
# Columnas a considerar para detectar duplicados parciales (todas excepto 'reservation_status_date')
columnas_para_parciales = [col for col in hoteles_limpio.columns if col != 'reservation_status_date']

# 1. Contar cuántas filas sobrantes hay con ese criterio (duplicados parciales)
# Usamos keep=False para identificar todas las instancias de los duplicados parciales
duplicados_parciales = hoteles_limpio[hoteles_limpio.duplicated(subset=columnas_para_parciales, keep=False)]
num_filas_parciales_sobrantes = duplicados_parciales.duplicated(subset=columnas_para_parciales, keep='first').sum()
print(f"Número de filas duplicadas parciales (idénticas excepto reservation_status_date): {num_filas_parciales_sobrantes}")

# 2. Mostrar algunas de esas filas ordenadas, con las columnas especificadas
columnas_a_mostrar = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'adr', 'reservation_status', 'reservation_status_date']

# Ordenar por las columnas de identificación para que los grupos parciales aparezcan juntos
duplicados_parciales_ordenados = duplicados_parciales.sort_values(by=columnas_para_parciales)

display(duplicados_parciales_ordenados[columnas_a_mostrar].head(10))

# 3. Mostrar la distribución de reservation_status en esas filas
print("\nDistribución de 'reservation_status' en las filas duplicadas parciales:")
display(duplicados_parciales['reservation_status'].value_counts(normalize=True))

Número de filas duplicadas parciales (idénticas excepto reservation_status_date): 245


,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adr,reservation_status,reservation_status_date
35572,City Hotel,9,2015,September,10,95.00,Canceled,2015-09-03
35574,City Hotel,9,2015,September,10,95.00,Canceled,2015-09-01
51267,City Hotel,12,2017,July,15,139.00,Canceled,2017-07-10
51268,City Hotel,12,2017,July,15,139.00,Canceled,2017-07-07
54170,City Hotel,14,2015,October,15,100.00,Canceled,2015-10-07
54474,City Hotel,14,2015,October,15,100.00,Canceled,2015-10-13
39425,City Hotel,14,2016,April,16,127.67,Canceled,2016-04-12
39432,City Hotel,14,2016,April,16,127.67,Canceled,2016-04-04
40080,City Hotel,14,2016,May,8,159.00,Canceled,2016-04-24
40083,City Hotel,14,2016,May,8,159.00,Canceled,2016-04-26



Distribución de 'reservation_status' en las filas duplicadas parciales:


,proportion
reservation_status,
Canceled,1.0


**Respuesta 4.2:**

_Cuando se cancela una reservación, se genera un segundo registro para la cancelación, es por eso que aparecen los mismo datos unos cuantos días despupes._

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

Respuesta 4.3:

Criterio de duplicado utilizado: Se consideraron todas las 32 columnas del dataset para la detección de duplicados exactos. Esto significa que dos filas se consideraron duplicadas si y solo si todos sus valores en cada una de las 32 columnas eran idénticos.

Regla de conservación: Se utilizó `keep='first'`, lo que implica que la primera aparición de cada grupo de filas duplicadas se mantuvo, y las instancias posteriores fueron eliminadas. Esta regla es relevante ya que se asumió que las copias exactas representan errores de registro, por lo que conservar una instancia es suficiente para mantener la información de la reserva.

Resultado:
- Filas antes de la limpieza: 119,390
- Filas después de la limpieza: 87,396
- Filas eliminadas: 31,994

Supuestos y limitaciones:
- Supuesto: Se asume que la existencia de filas idénticas en todas las 32 columnas representa un error en el proceso de registro de datos y no reservas distintas con detalles coincidentes. La ausencia de un identificador único de reserva refuerza esta interpretación, haciendo inviable distinguir entre una reserva real y su copia exacta.
- Limitación: La eliminación de duplicados exactos no aborda los duplicados parciales (identificados en la actividad 4.2), donde filas pueden ser idénticas en la mayoría de las columnas pero diferir en una columna como `reservation_status_date`. Estos duplicados parciales aún podrían influir en el análisis si representan el mismo evento con una actualización de estado.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

1. Una situación en la que NO convendría eliminar los duplicados exactos es si el problema de negocio dictara que cada fila, aunque idéntica en los datos disponibles, representa un evento o una intención de reserva independiente y válida. Por ejemplo, si el sistema de origen permite múltiples registros idénticos para rastrear intentos de reserva fallidos que luego se completan, y se desea analizar la frecuencia de esos intentos, aunque los datos finales sean los mismos.

2. Le pediría al proveedor un identificador único de reserva (`reservation_id`) y un identificador de cliente (`customer_id`). Con un `reservation_id`, se podría determinar de forma inequívoca si dos filas idénticas representan la misma reserva o dos reservas distintas que casualmente tienen los mismos detalles. Un `customer_id` ayudaría a entender si el mismo cliente está realizando múltiples reservas idénticas o si son clientes diferentes. Adicionalmente, una marca de tiempo de creación de la reserva (`booking_timestamp`) podría ser útil para diferenciar eventos idénticos creados en momentos distintos.

3. `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados porque solo detecta duplicados *exactos* en las columnas consideradas (por defecto, todas). Un dataset puede contener duplicados 'lógicos' o 'parciales', donde las filas son idénticas en un subconjunto clave de columnas, pero difieren en otras columnas no consideradas en la verificación, como fechas de actualización, IDs auxiliares o información menos relevante para el criterio de unicidad del negocio.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.